File: src-tauri/src/extraction/html.rs
use super::*;

pub fn strip_html(html: &str) -> String {
    let mut s = html.to_string();
    // CHG-28: when the page has a <main> or <article>, keep only that region so
    // the limited tool-result budget is spent on body text, not site chrome.
    for tag in ["<main", "<article"] {
        let lower = ascii_lower(&s);
        if let Some(open) = lower.find(tag) {
            let close = format!("</{}>", &tag[1..]);
            if let Some(rel) = lower.rfind(&close) {
                // A malformed page can put the closing tag before the opening
                // one; slicing backwards would panic, so leave `s` whole.
                if rel >= open {
                    s = s[open..rel + close.len()].to_string();
                    break;
                }
            }
        }
    }
    for tag in ["</p>", "</div>", "</li>", "</h1>", "</h2>", "</h3>", "</h4>", "</tr>", "<br>", "<br/>", "<br />"] {
        s = s.replace(tag, &format!("{tag}\n"));
    }
    // CHG-28: drop non-content element bodies (nav, chrome, forms, inline SVG)
    // in addition to scripts/styles, so their link text and boilerplate don't
    // crowd out the article.
    for pair in [
        ("<script", "</script>"),
        ("<style", "</style>"),
        ("<nav", "</nav>"),
        ("<header", "</header>"),
        ("<footer", "</footer>"),
        ("<aside", "</aside>"),
        ("<form", "</form>"),
        ("<noscript", "</noscript>"),
        ("<svg", "</svg>"),
    ] {
        loop {
            let lower = ascii_lower(&s);
            let Some(start) = lower.find(pair.0) else { break };
            match lower[start..].find(pair.1).map(|i| start + i + pair.1.len()) {
                Some(end) => s.replace_range(start..end, ""),
                None => break,
            }
        }
    }
    strip_tags(&s)
}

/// Case-folded copy for locating ASCII tag names, byte-for-byte aligned with
/// the input so an offset found here is a valid offset into the original.
///
/// `str::to_lowercase` is Unicode-aware and NOT length preserving — Turkish
/// `İ` (U+0130, 2 bytes) folds to `i` + U+0307 (3 bytes), shifting every later
/// offset. Applying those offsets to the original string sliced mid-character
/// and panicked the whole window mid-import. Tag names are ASCII, so ASCII
/// folding is both sufficient and safe.
fn ascii_lower(s: &str) -> String {
    s.to_ascii_lowercase()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn survives_length_changing_uppercase() {
        // Regression: `to_lowercase` expands Turkish `İ` (U+0130) from 2 bytes
        // to 3, so every offset past it pointed mid-character in the original.
        let html = "<div>İİİİ</div><main>İstanbul body</main><footer>İ chrome</footer>";
        let out = strip_html(html);
        assert!(out.contains("İstanbul body"), "got {out:?}");
        assert!(!out.contains("chrome"), "footer survived: {out:?}");
    }

    #[test]
    fn malformed_page_with_close_before_open_is_kept_whole() {
        // `</article>` before `<article` used to slice backwards and panic.
        let html = "</article><p>orphan text</p><article>tail";
        let out = strip_html(html);
        assert!(out.contains("orphan text"), "got {out:?}");
        assert!(out.contains("tail"), "got {out:?}");
    }

    #[test]
    fn drops_chrome_elements_case_insensitively() {
        let html = "<BODY><NAV>menu</NAV><p>keep me</p><SCRIPT>var x=1;</SCRIPT></BODY>";
        let out = strip_html(html);
        assert!(out.contains("keep me"), "got {out:?}");
        assert!(!out.contains("menu"), "got {out:?}");
        assert!(!out.contains("var x"), "got {out:?}");
    }
}

File: sidecar/arcelle_sidecar/docs/html.py
"""HTML text extraction: strip chrome/script/style elements, keep the main
article region when one exists, then strip remaining tags.

Port of `src-tauri/src/extraction/html.rs` (`strip_html`, `ascii_lower`).
Built on the shared `strip_tags` primitive in the sibling
`arcelle_sidecar.docs.xml_utils` module rather than redefining it.

----------------------------------------------------------------- ASCII fold

The Rust source searches for tag names in an ASCII-lowercased COPY of the
string, then slices/deletes from the ORIGINAL string using offsets found in
that copy. `str.to_lowercase()` (Rust) / `str.lower()` (Python) are both
Unicode-aware and NOT length-preserving -- Turkish `İ` (U+0130) folds to
`i` + a combining dot above (2 code points), which would shift every later
offset out of alignment with the original string, slicing mid-character or
raising. `_ascii_lower` below folds only `A`-`Z` to `a`-`z` and leaves every
other character (including non-ASCII ones) completely unchanged, so the
search copy stays EXACTLY the same length -- and character-for-character
aligned with the original at every index -- no matter what it contains. Tag
names are always ASCII, so this loses nothing.
"""

from __future__ import annotations

from arcelle_sidecar.docs.xml_utils import strip_tags

# CHG-28: when the page has a <main> or <article>, keep only that region so
# the limited tool-result budget is spent on body text, not site chrome.
_REGION_TAGS: tuple[str, ...] = ("<main", "<article")

# Paragraph/line-break-ish closers: a literal newline is appended after each
# occurrence so text doesn't run together once tags are stripped. Matched
# CASE-SENSITIVELY, exactly as the Rust source does -- this pass runs before
# any lowercasing.
_BREAK_TAGS: tuple[str, ...] = (
    "</p>",
    "</div>",
    "</li>",
    "</h1>",
    "</h2>",
    "</h3>",
    "</h4>",
    "</tr>",
    "<br>",
    "<br/>",
    "<br />",
)

# CHG-28: drop non-content element bodies (nav, chrome, forms, inline SVG)
# in addition to scripts/styles, so their link text and boilerplate don't
# crowd out the article.
_CHROME_PAIRS: tuple[tuple[str, str], ...] = (
    ("<script", "</script>"),
    ("<style", "</style>"),
    ("<nav", "</nav>"),
    ("<header", "</header>"),
    ("<footer", "</footer>"),
    ("<aside", "</aside>"),
    ("<form", "</form>"),
    ("<noscript", "</noscript>"),
    ("<svg", "</svg>"),
)


def _ascii_lower(s: str) -> str:
    """`str::to_ascii_lowercase` -- fold only 'A'-'Z', leave every other
    character (including non-ASCII ones) untouched, so the result stays
    exactly as long as `s` and aligned with it index-for-index. Deliberately
    NOT `str.lower()` (Unicode-aware, can change length -- see module
    docstring).
    """
    return "".join(chr(ord(c) + 32) if "A" <= c <= "Z" else c for c in s)


def strip_html(html: str) -> str:
    s = html
    for tag in _REGION_TAGS:
        lower = _ascii_lower(s)
        open_pos = lower.find(tag)
        if open_pos == -1:
            continue
        close = f"</{tag[1:]}>"
        rel = lower.rfind(close)
        if rel == -1:
            continue
        # A malformed page can put the closing tag before the opening one;
        # slicing backwards would be wrong, so leave `s` whole.
        if rel >= open_pos:
            s = s[open_pos : rel + len(close)]
            break

    for tag in _BREAK_TAGS:
        s = s.replace(tag, f"{tag}\n")

    for open_tag, close_tag in _CHROME_PAIRS:
        while True:
            lower = _ascii_lower(s)
            start = lower.find(open_tag)
            if start == -1:
                break
            found = lower.find(close_tag, start)
            if found == -1:
                break
            end = found + len(close_tag)
            s = s[:start] + s[end:]

    return strip_tags(s)
